rotkehlchen/chain/ethereum/modules/balancer/utils.py

```diff
@@ -0,0 +1,299 @@
+            token0_name = raw_token0['name']
+            token0_decimals = raw_token0['decimals']
+            token1_name = raw_token1['name']
+            token1_decimals = raw_token1['decimals']
+        except KeyError as e:
+            raise DeserializationError(f'Missing key: {str(e)}.') from e
+    else:
+        token0_name = None
+        token0_decimals = None
+        token1_name = None
+        token1_decimals = None
+
+    token0 = get_ethereum_token(
+        symbol=token0_symbol,
+        ethereum_address=addresses.token_in,
+        name=token0_name,
+        decimals=token0_decimals,
+    )
+    token1 = get_ethereum_token(
+        symbol=token1_symbol,
+        ethereum_address=addresses.token_out,
+        name=token1_name,
+        decimals=token1_decimals,
+    )
+    amm_swap = AMMSwap(
+        tx_hash=tx_hash,
+        log_index=log_index,
+        address=addresses.user_address,
+        from_address=addresses.caller,
+        to_address=addresses.pool_address,
+        timestamp=timestamp,
+        location=Location.BALANCER,
+        token0=token0,
+        token1=token1,
+        amount0_in=amount0_in,
+        amount1_in=AssetAmount(ZERO),
+        amount0_out=AssetAmount(ZERO),
+        amount1_out=amount1_out,
+    )
+    return amm_swap
+
+
+def deserialize_token_price(
+        raw_token_price: Dict[str, Any],
+) -> Tuple[ChecksumEthAddress, Price]:
+    """May raise DeserializationError"""
+    try:
+        token_address = raw_token_price['id']
+        usd_price = deserialize_price(raw_token_price['price'])
+    except KeyError as e:
+        raise DeserializationError(f'Missing key: {str(e)}.') from e
+
+    try:
+        token_address = to_checksum_address(token_address)
+    except ValueError as e:
+        raise DeserializationError(
+            f'Invalid ethereum address: {token_address} in token price.',
+        ) from e
+
+    return token_address, usd_price
+
+
+def deserialize_token_day_data(
+        raw_token_day_data: Dict[str, Any],
+) -> Tuple[ChecksumEthAddress, Price]:
+    """May raise DeserializationError"""
+    try:
+        token_address = raw_token_day_data['token']['id']
+        usd_price = deserialize_price(raw_token_day_data['priceUSD'])
+    except KeyError as e:
+        raise DeserializationError(f'Missing key: {str(e)}.') from e
+
+    try:
+        token_address = to_checksum_address(token_address)
+    except ValueError as e:
+        raise DeserializationError(
+            f'Invalid ethereum address: {token_address} in token day data.',
+        ) from e
+
+    return token_address, usd_price
+
+
+def calculate_amm_trade_from_amm_swaps(swaps: List[AMMSwap]) -> AMMTrade:
+    """AMMTrade is always a buy.
+
+    [USDC -> AMPL]               BASE_QUOTE pair is AMPL_USDC.
+    [USDC -> AMPL, AMPL -> WETH] BASE_QUOTE pair is WETH_USDC.
+    """
+    if len(swaps) == 0:
+        raise AssertionError("Swaps can't be an empty list")
+
+    amm_trade = AMMTrade(
```

Comment:
It can be more than 2 swaps. For instance this trade done via the proxy has 3 swaps: wETH -> DAI -> HOPR
https://etherscan.io/tx/0x730149f5e331951e330880ff72bdab326c7b8bdf7e7ae0e57a9abff6fbca3103

The trade will be wETH to HOPR, but it will contain the 3 swaps.

Reply:
So you are only taking the first and last of a list of at least 1 trades? But can also be more? But you ignore the in-between?

Reply:
Yes, is it wrong? The quantities match, the account used the proxy for swapping wETH to HOPR, that was its "trade", what happened in the middle are swaps (that we save within the structure and the db) but as far as I understand do not reflect the trade intention. This can be seen using their UI (the swap routes I mean).

Reply:
Uniswap trades/swaps follow the same pattern, we have tests for that as well.

Reply:
No I am not saying it's wrong. I just don't understand what you are doing by just reading the code. Please mention somewhere in the docstring why you take first and last and also that for the purposes of this function the in-between are ignored